blockware/matrix/lib/MatrixRain/MatrixRain.cpp

```diff
@@ -0,0 +1,225 @@
+#include <Adafruit_SSD1351.h>
+#include <math.h>
+
+#include <Colors.h>
+#include <Random.h>
+
+#include "MatrixRain.h"
+
+const char* UNKNOWN = "?";
+const float RESTART_CHANCE = 0.75;
+const float WHITE_RANDOM_CHANGE = 0.20;
+const float GREEN_RANDOM_CHANGE = 0.01;
+const uint8_t VISIBLE_DARKNESS = 0;
+const uint8_t NOT_VISIBLE_DARKNESS = 255;
+const uint8_t MIN_WHITE_THRESHOLD = NOT_VISIBLE_DARKNESS / 32;
+const uint8_t MAX_WHITE_THRESHOLD = NOT_VISIBLE_DARKNESS / 8;
+const uint8_t MIN_DARKNESS_RATE = 4;
+const uint8_t MAX_DARKNESS_RATE = 8;
+
+static const char* alphabet[] PROGMEM = {
+  // Symbols
+  "!", "?", "@", "#", "$", "%", "&", "*", "+", "-", ":", "<", ">", ";",
+  // Numbers
+  "0",  "1",  "2",  "3",  "4",  "5", "6",  "7",  "8",  "9",
+  // Latin
+  "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
+  "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z",
+  // Halfwidth Katakana
+  "ｦ", "ｧ", "ｨ", "ｩ", "ｪ", "ｫ", "ｬ", "ｭ", "ｮ", "ｯ",
+  "ｰ", "ｱ", "ｲ", "ｳ", "ｴ", "ｵ", "ｶ", "ｷ", "ｸ", "ｹ", "ｺ", "ｻ", "ｼ", "ｽ", "ｾ", "ｿ",
+  "ﾀ", "ﾁ", "ﾂ", "ﾃ", "ﾄ", "ﾅ", "ﾆ", "ﾇ", "ﾈ", "ﾉ", "ﾊ", "ﾋ", "ﾌ", "ﾍ", "ﾎ", "ﾏ",
+  "ﾐ", "ﾑ", "ﾒ", "ﾓ", "ﾔ", "ﾕ", "ﾖ", "ﾗ", "ﾘ", "ﾙ", "ﾚ", "ﾛ", "ﾜ", "ﾝ",
+};
+
+static const char* hiragana[] PROGMEM = {
+  // Hiragana
+  // https://www.unicode.org/charts/PDF/U3040.pdf
+  "ぁ", "あ", "ぃ", "い", "ぅ", "う", "ぇ", "え", "ぉ", "お", "か", "が", "き",
+  "ぎ", "く", "ぐ", "け", "げ", "こ", "ご", "さ", "ざ", "し", "じ", "す", "ず",
+  "せ", "ぜ", "そ", "ぞ", "た", "だ", "ち", "ぢ", "っ", "つ", "づ", "て", "で",
+  "と", "ど", "な", "に", "ぬ", "ね", "の", "は", "ば", "ぱ", "ひ", "び", "ぴ",
+  "ふ", "ぶ", "ぷ", "へ", "べ", "ぺ", "ほ", "ぼ", "ぽ", "ま", "み", "む", "め",
+  "も", "ゃ", "や", "ゅ", "ゆ", "ょ", "よ", "ら", "り", "る", "れ", "ろ", "ゎ",
+  "わ", "ゐ", "ゑ", "を", "ん", "ゔ", "ゕ", "ゖ", "ゝ", "ゞ", "ゟ",
+};
+
+// find length of array using pointer arithmetic
+// since we have a pointer at the start of the array, the length of the​ array can be calculated
+// let A = arr        the address of the start of the array
+// let B = &arr + 1   points to the memory address right after the end of the array
+// let C = *B         casts the above address to an int *
+// subtracting A from C gives the length of the array
+// int length = *(&arr + 1) - arr;
+int alphabet_length = *(&alphabet + 1) - alphabet;
+int hiragana_length = *(&hiragana + 1) - hiragana;
+
+Vec2d<uint8_t> alphabet_textSize = Vec2d<uint8_t>(8, 12);
+Vec2d<uint8_t> hiragana_textSize = Vec2d<uint8_t>(12, 14);
+
+Vec2d<uint8_t> MatrixRain::cell = Vec2d<uint8_t>(std::max(hiragana_textSize.x, alphabet_textSize.x), std::max(hiragana_textSize.y, alphabet_textSize.y));
+
+Vec2d<uint8_t> centerCellVec2d(Vec2d<uint8_t>& relativeVector) {
+  return Vec2d<uint8_t>((MatrixRain::cell.x / 2) - (relativeVector.x / 2), (MatrixRain::cell.y / 2) - (relativeVector.y / 2));
+}
+
+uint8_t rand_white_darkness() {
+  return (uint8_t)time_random(MIN_WHITE_THRESHOLD, MAX_WHITE_THRESHOLD);
+}
+
+uint8_t rand_darkness_rate() {
```

Comment:
Nit: Most functions, methods, and variables in our blockware (including the other functions in this file) use `camelCase`

Perhaps I'll set up clang-tidy to enforce this automatically so we don't have to think about it.

Reply:
But if this is forked from something else, fine to leave in the original style so it would be easy to update if needed.

Reply:
oh i just felt it was easier to read this way so things were consistently cased when they are derived like `letter` and `letter_vec`, etc. it also makes it easier to refactor later w find replace. i totally get it tho o can take a pass and see if i can fix this another way and keep camel case